File: backend/app/models/project.py
from datetime import datetime, timezone, date
from sqlalchemy import String, Text, DateTime, Date, ForeignKey, Enum as SAEnum, Integer
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base
from app.utils.enums import ProjectStatus, Priority


class Project(Base):
    __tablename__ = "projects"

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    company_id: Mapped[int] = mapped_column(ForeignKey("companies.id", ondelete="RESTRICT"), nullable=False)
    department_id: Mapped[int | None] = mapped_column(ForeignKey("departments.id", ondelete="SET NULL"))
    main_responsible_id: Mapped[int | None] = mapped_column(ForeignKey("users.id", ondelete="SET NULL"))
    name: Mapped[str] = mapped_column(String(250), nullable=False, index=True)
    description: Mapped[str | None] = mapped_column(Text)
    start_date: Mapped[date | None] = mapped_column(Date)
    deadline: Mapped[date | None] = mapped_column(Date)
    status: Mapped[ProjectStatus] = mapped_column(
        SAEnum(ProjectStatus, name="project_status", values_callable=lambda x: [e.value for e in x]),
        default=ProjectStatus.PLANNED,
        nullable=False,
    )
    priority: Mapped[Priority] = mapped_column(
        SAEnum(Priority, name="priority_level", values_callable=lambda x: [e.value for e in x]),
        default=Priority.MEDIUM,
        nullable=False,
    )
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(timezone.utc)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(timezone.utc),
        onupdate=lambda: datetime.now(timezone.utc),
    )

    # Relaciones
    company: Mapped["Company"] = relationship("Company", back_populates="projects")
    department: Mapped["Department | None"] = relationship("Department")
    main_responsible: Mapped["User | None"] = relationship("User", foreign_keys=[main_responsible_id])
    activities: Mapped[list["Activity"]] = relationship(
        "Activity", back_populates="project", cascade="all, delete-orphan"
    )

File: mobile/lib/models/project.dart
import 'package:flutter/foundation.dart';

class Project {
  final int id;
  final int companyId;
  final String? companyName;
  final String name;
  final String? description;
  final String? startDate;
  final String? deadline;
  final String status;
  final String priority;
  final int? mainResponsibleId;
  final String? mainResponsibleName;
  final String createdAt;

  Project({
    required this.id,
    required this.companyId,
    this.companyName,
    required this.name,
    this.description,
    this.startDate,
    this.deadline,
    this.status = 'planificado',
    this.priority = 'media',
    this.mainResponsibleId,
    this.mainResponsibleName,
    required this.createdAt,
  });

  factory Project.fromJson(Map<String, dynamic> json) {
    return Project(
      id: json['id'],
      companyId: json['company_id'],
      companyName: json['company_name'],
      name: json['name'],
      description: json['description'],
      startDate: json['start_date'],
      deadline: json['deadline'],
      status: json['status'] ?? 'planificado',
      priority: json['priority'] ?? 'media',
      mainResponsibleId: json['main_responsible_id'],
      mainResponsibleName: json['main_responsible_name'],
      createdAt: json['created_at'] ?? '',
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'company_id': companyId,
      'name': name,
      'description': description,
      'start_date': startDate,
      'deadline': deadline,
      'status': status,
      'priority': priority,
      'main_responsible_id': mainResponsibleId,
    };
  }
}
